tools/not.cpp

```diff
@@ -0,0 +1,76 @@
+//===- not.cpp - The 'not' testing tool -----------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Usage:
+//   not cmd
+//     Will return true if cmd doesn't crash and returns false.
+//   not --crash cmd
+//     Will return true if cmd crashes (e.g. for testing crash reporting).
+
+// This file is a stripped down version of not.cpp from llvm/utils. This does
+// not depend on any LLVM library.
+
+#include <cstdlib>
+#include <iostream>
+#include <sstream>
+
+#ifdef _WIN32
+#include <windows.h>
+#endif
+
+int main(int argc, const char **argv) {
+  bool expectCrash = false;
+
+  ++argv;
+  --argc;
+
+  if (argc > 0 && std::string(argv[0]) == "--crash") {
+    ++argv;
+    --argc;
+    expectCrash = true;
+
+    // Crash is expected, so disable crash report and symbolization to reduce
+    // output and avoid potentially slow symbolization.
+#ifdef _WIN32
+    SetEnvironmentVariableA("LLVM_DISABLE_CRASH_REPORT", "1");
+    SetEnvironmentVariableA("LLVM_DISABLE_SYMBOLIZATION", "1");
+#else
+    setenv("LLVM_DISABLE_CRASH_REPORT", "1", 0);
+    setenv("LLVM_DISABLE_SYMBOLIZATION", "1", 0);
+#endif
+  }
+
+  if (argc == 0)
+    return 1;
+
+  std::stringstream ss;
+  ss << argv[0];
+  for (int i = 1; i < argc; ++i)
+    ss << " " << argv[i];
+  std::string cmd = ss.str();
+  int result = system(cmd.c_str());
+
+#ifdef _WIN32
+  // Handle abort() in msvcrt -- It has exit code as 3.  abort(), aka
+  // unreachable, should be recognized as a crash.  However, some binaries use
+  // exit code 3 on non-crash failure paths, so only do this if we expect a
+  // crash.
+  if (expectCrash && result == 3)
+    result = -3;
+#endif
+
+  if (result < 0) {
+    if (expectCrash)
+      return 0;
+    return 1;
+  }
+
+  if (expectCrash)
+    return 1;
+
+  return result == 0;
```

Comment:
Doesn't 0 mean successful termination on unix-like systems? In which case, wouldn't this not invert the return status of the program?

Maybe this would be clearer if you used the EXIT_SUCCESS constant from https://en.cppreference.com/w/c/program/EXIT_status

Reply:
Oh I understand now - when result is 0, the condition evaluates to true (1) and so this inverts the error code. Nevermind.

Reply:
I started to respond and then got stuck thinking more closely about what this code is doing, and there are some issues here too. The original code uses `llvm::sys::ExecuteAndWait` which returns -2 on an abort or a timeout. The `result < 0` condition checks this. But I am not sure if that is the case with the libc `system` call that I have used. I will try to find out, at least on Linux, and add comments to the code accordingly.

Thanks for raising the question :)

Reply:
system does not directly returns exit_code, at least not on Linux. (It returns a status code and be then interpreted as exit snd stopsig code) There is a macro can be used to get exit code from status: https://en.cppreference.com/w/cpp/utility/program/system
On POSIX systems, the return value can be decomposed using [WEXITSTATUS and WSTOPSIG](https://pubs.opengroup.org/onlinepubs/9699919799/functions/wait.html).
If I remember correctly the WEXITSTATUS calculate exit code by and 0xFF with a 8 shift left? 
This does not apply on Windows, I am not sure if the return code from std::system matches the Windows error code or not https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-
TL;DR: so basically a value of 0 means no problem on Windows and Linux

Reply:
Thanks @tblah and @yi-wu-arm. 

I have fixed the code to check for a signal that causes termination and added some comments explaining what is going on.